### ISSUE TITLE
systemd-networkd-wait-online times out, with a bond interface and two slaves

### ISSUE THREAD

**Issue by manuthambi** [systemd/systemd#29506]
### systemd version the issue has been seen with

252.17

### Used distribution

Debian 12

### Linux kernel version used

6.1.0-12-amd64

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd-wait-online

### Expected behaviour you didn't see

I expect `systemd-networkd-wait-online` to exit once one of the enslaved interfaces of `bond0` is connected and online. 
This used to work fine with systemd version `252.12`

### Unexpected behaviour you saw

`systemd-networkd-wait-online` waits until it times out.

### Steps to reproduce the problem

I have a `bond0` interface with two slaves `en_lan0` and `en_lan1`. `en_lan0` is not connected, and `en_lan1` is connected. Both slaves have `ConfigureWithoutCarrier=yes`.  `systemd-networkd-wait-online`  should exit with success once the `bond0` is online, and both slaves are configured.

The `man` page has: `By default, it will wait for all links it is aware of and which are managed by systemd-networkd.service(8) to be fully configured or failed, and for at least one link to be online.`

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
$ cat /etc/systemd/network/10-en_lan0.link 
[Match]
MACAddress =  00:25:90:7a:16:d0
OriginalName = eth*

[Link]
Name = en_lan0
```
```sh
$ cat /etc/systemd/network/10-en_lan0.network 
[Match]
Name=en_lan0

[Network]
Bond=bond0
ConfigureWithoutCarrier=yes
```
```sh
$ cat /etc/systemd/network/10-en_lan0.link   
[Match]
MACAddress =  00:25:90:7a:16:d0
OriginalName = eth*
```
```sh
$ cat /etc/systemd/network/10-en_lan1.network
[Match]
Name=en_lan1

[Network]
Bond=bond0
ConfigureWithoutCarrier=yes
```
```sh
$ cat /etc/systemd/network/10-bond0.netdev 
[NetDev]
Name=bond0
Kind=bond

[Bond]
Mode=active-backup
MIIMonitorSec=100ms
```
```sh
$ cat /etc/systemd/network/10-bond0.network
[Match]
Name=bond0

[Network]
Address=172.23.23.44/24
Gateway=172.23.23.1
```
```sh
$ networkctl status bond0
● 6: bond0                                                                     
                     Link File: /usr/lib/systemd/network/99-default.link
                  Network File: /etc/systemd/network/10-bond0.network
                         State: routable (configured)
                  Online state: online                                         
                          Type: bond
                          Kind: bond
                        Driver: bonding
              Hardware Address: b2:a1:7c:d4:32:40
                           MTU: 1500 (min: 68, max: 65535)
                         QDisc: noqueue
  IPv6 Address Generation Mode: eui64
                          Mode: active-backup
                        Miimon: 100ms
                       Updelay: 0
                     Downdelay: 0
      Number of Queues (Tx/Rx): 16/16
              Auto negotiation: no
                         Speed: 1Gbps
                        Duplex: full
                       Address: 172.23.23.44
                                fe80::b0a1:7cff:fed4:3240
                       Gateway: 172.23.23.1
                           DNS: 172.23.23.1
                Search Domains: office.m
                 Route Domains: m
                                in.meshcapital.com
                                23.172.in-addr.arpa
                                27.172.in-addr.arpa
             Activation Policy: up
           Required For Online: yes
             DHCP6 Client DUID: DUID-EN/Vendor:0000ab1101627cc5247bf9d70000

Oct 09 06:19:20 compute4.office.m systemd-networkd[430]: bond0: DHCPv6 lease lost
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: bond0: netdev ready
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: bond0: Link UP
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: bond0: Gained carrier
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: bond0: Gained IPv6LL
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: bond0: netdev exists, using existing without changing its parameters
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: bond0: Configuring with /etc/systemd/network/10-bond0.network.
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: bond0: Lost carrier
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: bond0: DHCPv6 lease lost
Oct 09 06:19:24 compute4.office.m systemd-networkd[48713]: bond0: Gained carrier
```
```sh
$ networkctl status en_lan0
● 2: en_lan0                                                                  
                     Link File: /usr/lib/systemd/network/99-default.link
                  Network File: /etc/systemd/network/10-en_lan0.network
                         State: no-carrier (configured)
                  Online state: offline
                          Type: ether
                          Path: pci-0000:04:00.0
                        Driver: igb
                        Vendor: Intel Corporation
                         Model: I350 Gigabit Network Connection
             Alternative Names: enp4s0f0
              Hardware Address: b2:a1:7c:d4:32:40
    Permanent Hardware Address: 00:25:90:7a:16:d0 (Super Micro Computer, Inc.)
                           MTU: 1500 (min: 68, max: 9216)
                         QDisc: mq
                        Master: bond0
  IPv6 Address Generation Mode: none
      Number of Queues (Tx/Rx): 8/8
              Auto negotiation: yes
                          Port: tp
             Activation Policy: up
           Required For Online: yes

Sep 26 04:14:29 compute4.office.m systemd-networkd[430]: en_lan0: Configuring with /etc/systemd/network/10-en_lan0.network.
Sep 26 04:14:29 compute4.office.m systemd-networkd[430]: en_lan0: Link UP
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: en_lan0: Link UP
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: en_lan0: Configuring with /etc/systemd/network/10-en_lan0.network.
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: en_lan0: Link DOWN
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: en_lan0: Link UP
```
```sh
$ networkctl status en_lan1
● 3: en_lan1                                                                   
                     Link File: /usr/lib/systemd/network/99-default.link
                  Network File: /etc/systemd/network/10-en_lan1.network
                         State: enslaved (configured)
                  Online state: online                                         
                          Type: ether
                          Path: pci-0000:04:00.1
                        Driver: igb
                        Vendor: Intel Corporation
                         Model: I350 Gigabit Network Connection
             Alternative Names: enp4s0f1
              Hardware Address: b2:a1:7c:d4:32:40
    Permanent Hardware Address: 00:25:90:7a:16:d1 (Super Micro Computer, Inc.)
                           MTU: 1500 (min: 68, max: 9216)
                         QDisc: mq
                        Master: bond0
  IPv6 Address Generation Mode: none
      Number of Queues (Tx/Rx): 8/8
              Auto negotiation: yes
                         Speed: 1Gbps
                        Duplex: full
                          Port: tp
             Activation Policy: up
           Required For Online: yes

Sep 26 04:14:29 compute4.office.m systemd-networkd[430]: en_lan1: Configuring with /etc/systemd/network/10-en_lan1.network.
Sep 26 04:14:29 compute4.office.m systemd-networkd[430]: en_lan1: Link UP
Sep 26 04:14:32 compute4.office.m systemd-networkd[430]: en_lan1: Gained carrier
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: en_lan1: Link UP
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: en_lan1: Gained carrier
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: en_lan1: Configuring with /etc/systemd/network/10-en_lan1.network.
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: en_lan1: Link DOWN
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: en_lan1: Lost carrier
Oct 09 06:19:20 compute4.office.m systemd-networkd[48713]: en_lan1: Link UP
Oct 09 06:19:24 compute4.office.m systemd-networkd[48713]: en_lan1: Gained carrier
```
```sh
$ SYSTEMD_LOG_LEVEL=debug /lib/systemd/systemd-networkd-wait-online --timeout=5; echo $?
Found link 6
Found link 5
Found link 4
Found link 3
Found link 2
Found link 1
en_lan1: link is configured by networkd and online.
bond0: link is configured by networkd and online.
enp3s0d1: link is not managed by networkd.
lo: link is ignored
enp3s0: link is not managed by networkd.
en_lan0: Operational state 'no-carrier' is not in range ['degraded':'routable']
Timeout occurred while waiting for network connectivity.
1
```

**Comment by yuwata** [comment]
Hmm... Actually ab3aed4a0349bbaa26f53340770c1b59b463e05d (which is backported to v252-stable in v252.15) changes the behavior...

**Comment by yuwata** [comment]
Sorry for late. Fix is waiting in #31184.
